File: 백준/Gold/5427. 불/불.py
from collections import deque
import sys
input = sys.stdin.readline

T = int(input())

for _ in range(T):
    w, h = map(int, input().split())
    building = [list(input().strip()) for _ in range(h)]
    fire_time = [[-1] * w for _ in range(h)]
    visited = [[False] * w for _ in range(h)]
    dy = [-1, 1, 0, 0]
    dx = [0, 0, -1, 1]
    fire = deque()
    person = deque()

    for y in range(h):
        for x in range(w):
            if building[y][x] == '*':
                fire.append((y, x))
                fire_time[y][x] = 0
            elif building[y][x] == '@':
                person.append((y, x, 0))
                visited[y][x] = True

    while fire:
        y, x = fire.popleft()
        for d in range(4):
            ny = y + dy[d]
            nx = x + dx[d]
            if 0 <= ny < h and 0 <= nx < w:
                if building[ny][nx] == '.' and fire_time[ny][nx] == -1:
                    fire_time[ny][nx] = fire_time[y][x] + 1
                    fire.append((ny, nx))

    escaped = False
    while person:
        y, x, t = person.popleft()
        for d in range(4):
            ny = y + dy[d]
            nx = x + dx[d]
            if not (0 <= ny < h and 0 <= nx < w):
                print(t + 1)
                escaped = True
                break
            if building[ny][nx] != '.' or visited[ny][nx]:
                continue
            if fire_time[ny][nx] != -1 and fire_time[ny][nx] <= t + 1:
                continue
            visited[ny][nx] = True
            person.append((ny, nx, t + 1))
        if escaped:
            break

    if not escaped:
        print("IMPOSSIBLE")

File: 백준/Gold/5427. 불/불.js
const fs = require("fs");
const path = require("path");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

let idx = 0;
let T = parseInt(input[idx++]);

const dx = [0, 0, -1, 1];
const dy = [-1, 1, 0, 0];

while (T--) {
  const [w, h] = input[idx++].split(" ").map(Number);
  const building = [];
  const fireTime = Array.from({ length: h }, () => Array(w).fill(-1));
  const visited = Array.from({ length: h }, () => Array(w).fill(false));
  const fire = [];
  const person = [];

  for (let i = 0; i < h; i++) {
    const row = input[idx++].split("");
    building.push(row);
    for (let j = 0; j < w; j++) {
      if (row[j] === "*") {
        fire.push([i, j]);
        fireTime[i][j] = 0;
      } else if (row[j] === "@") {
        person.push([i, j, 0]);
        visited[i][j] = true;
      }
    }
  }

  let fIdx = 0;
  while (fIdx < fire.length) {
    const [y, x] = fire[fIdx++];
    for (let d = 0; d < 4; d++) {
      const ny = y + dy[d];
      const nx = x + dx[d];
      if (ny >= 0 && ny < h && nx >= 0 && nx < w) {
        if (building[ny][nx] === "." && fireTime[ny][nx] === -1) {
          fireTime[ny][nx] = fireTime[y][x] + 1;
          fire.push([ny, nx]);
        }
      }
    }
  }

  let pIdx = 0;
  let escaped = false;

  while (pIdx < person.length) {
    const [y, x, t] = person[pIdx++];
    for (let d = 0; d < 4; d++) {
      const ny = y + dy[d];
      const nx = x + dx[d];

      if (ny < 0 || ny >= h || nx < 0 || nx >= w) {
        console.log(t + 1);
        escaped = true;
        break;
      }

      if (building[ny][nx] !== "." || visited[ny][nx]) continue;
      if (fireTime[ny][nx] !== -1 && fireTime[ny][nx] <= t + 1) continue;

      visited[ny][nx] = true;
      person.push([ny, nx, t + 1]);
    }
    if (escaped) break;
  }

  if (!escaped) console.log("IMPOSSIBLE");
}
